source/extensions/common/wasm/plugin.h

```diff
@@ -46,7 +48,7 @@ class Plugin : public proxy_wasm::PluginBase {
                    config.vm_config().runtime(), MessageUtil::anyToBytes(config.configuration()),
                    config.fail_open(), createPluginKey(config, direction, listener_metadata)),
         direction_(direction), local_info_(local_info), listener_metadata_(listener_metadata),
-        wasm_config_(std::make_unique<WasmConfig>(config)) {}
+        wasm_config_(std::make_unique<WasmConfig>(config, Stats::getCustomStatNamespaces())) {}
```

Comment:
can we pass the `CustomStatNamespaces&` into the Plugin? 

I'm going to keep asking that question until we no longer need a mutable static, or we reach a point where it really can't be done for some reason.

Reply:
Possible, but beyond this that would cost a lot more code change  - we need to change all the Wasm extension points (http, network, access logger, stat sink, and wasm service). I don't think that makes to export this shared logic to every one of these extensions.

Reply:
cc @PiotrSikora

Reply:
How much code to change?

I think if we have to change ~6 code-points in wasm extension that's obviously a win.

I think we have to change all the *other* extensions then it would be beyond the scope of this PR.

WDYT?

Reply:
More specifically, IIRC, extensions are expected to get an API object (but don't know the details). I think it would make sense to add a CustomStatNameSpace to the API object.

Reply:
> extensions are expected to get an API object (but don't know the details). I think it would make sense to add a CustomStatNameSpace to the API object.

thanks, that really makes sense.. will do follow-up soon!

Reply:
Finally everything you mentioned as "no new static global is needed" makes sense to me. In  https://github.com/envoyproxy/envoy/pull/17357/commits/63a74265f8e110e9f53b930ba1f929eae85fb75e, we no longer need to have static new singleton stuff, and instead we embed it as a member of API object, and all call sites can retrieve from the API object they receive. the much clearer and the less hassle in tests. Thank you so much!